Move the `tsh` config file guidance (#30437)
gravitational/teleport: docs/pages/connect-your-client/tsh.mdx

aliases:
    "<alias>": "<command>"
```

The `<alias>` can only be a top-level subcommand. In other words, you can define a `tsh mycommand` alias but not `tsh my command`.

New command `tsh l`:

```yaml
aliases:
    "l": "tsh login --auth=okta"
```

Make `tsh status` use JSON as a default format:

```yaml
aliases:
    "status": "tsh status --format=json"
```

The alias can use an arbitrary number of arguments. If an argument variable `$N` is referenced, `tsh` will check that at least `N+1` arguments were given to the alias invocation. All arguments that were given but not referenced in the alias definition will be appended at the end.

Define a custom command using `bash`. The `$0` and `$1` variables will be substituted with command arguments.

```yaml
aliases:
    "connect": "bash -c 'tsh login $0 && tsh ssh $1'"
```

Define a custom login command where first argument specifies `--auth` option.

```yaml
aliases:
    "ap": "tsh login --auth=$0 --proxy=teleport.example.com"
```

Given the configuration:

```yaml
aliases:
    "example": "bash -c 'echo first=$0 $0-$1 $3'"
```

`tsh example 0 1 unused-2 3 unused-4` will expand to `bash -c 'echo first=0 0-1 3 unused-2 unused-4'`.

> An alias definition can also reference the `$TSH` variable. `tsh` will expand this variable to the absolute path of the current `tsh` executable. This can be useful if there are multiple `tsh` versions installed, or the currently used version is not in `PATH`.

Reviewer: ```suggestion
An alias definition can also reference the `$TSH` variable. If you use the $TSH variable in an alias, `tsh` expands the variable to the absolute path of the current `tsh` executable. This behavior can be useful if there are multiple `tsh` versions installed, or the version you're currently using is not in the `PATH`.
```

```yaml
aliases:
    "status": "$TSH status --format=json"
```

> To troubleshoot aliases, set the `TELEPORT_DEBUG=1` environment variable. This will cause detailed logs to be printed to the console:

Reviewer: ```suggestion
To troubleshoot aliases, set the `TELEPORT_DEBUG=1` environment variable. With this variable, detailed logs are displayed as standard output. For example:
```

Author: 👍 ,though this should be "standard error". I'm also not sure what "With this variable" adds, so I've preserved "This".

```bash
$ TELEPORT_DEBUG=1 tsh status
DEBU [TSH]       Self re-exec command: tsh [status --format=json]. tsh/aliases.go:203
...
```

### Proxy Templates

> With Proxy Templates, `tsh` will dynamically determine the address of the
> Proxy Service to connect to based on the address of the destination host
> in your `tsh ssh` or `tsh proxy ssh` command.

Reviewer: ```suggestion
With proxy templates, `tsh` dynamically determines the address of the
Teleport Proxy Service to connect to based on the address of the destination host
in your `tsh ssh` or `tsh proxy ssh` command.
```

```yaml
proxy_templates:

# Regular expression that the host server address `%h:%p` is matched against.
# The "replace rules" below can reference capturing groups from this regular
# expression (`$1`, `$2`, etc.).
- template: '^(\w+)\.(\w+):([0-9]+)$' # <nodename>.<clustername>:<port>

  # Optional web proxy address to use for proxy jump (`--jumphost`, `-J`).
  #
  # Proxy Jump can be used to reduce latency in regionally distributed trusted
  # clusters by connecting to a leaf node through the leaf proxy instead of the
  # root proxy.
  proxy: "$2.eu.example.com:443"

  # Optional cluster name to connect to (`--cluster`).
  #
  # Cluster can be used to connect to leaf nodes from the root proxy without
  # first logging in to the leaf cluster. This may be useful in cases where
  # proxy jump is not applicable, such as when the leaf clusters do not have
  # their own public proxies.
  cluster: "$2"

  # Optional host server address to connect to (`%h:%p`).
  #
  # Port defaults to 3022 if not explicitly provided with `--port`.
  host: "$1:$3"

# Multiple templates can be provided. They are evaluated in order and the first

> # one matching will take effect.

Reviewer: ```suggestion
# match takes effect.
```

- template: ...
```

`tsh -J {{proxy}} ssh` and `tsh -J {{proxy}} proxy ssh` will attempt to match the
host server address `%h:%p` with the configured templates. For each replace rule set,
the corresponding cli value will be set.

> If leaf certificates are required to connect to the node, `tsh` will automatically
> retrieve leaf certificates from the root cluster.

Reviewer: ```suggestion
If leaf certificates are required to connect to the node, `tsh` automatically
retrieves leaf certificates from the root cluster.
```

```bash
$ tsh ssh -J {{proxy}} node1.leaf1
# becomes
$ tsh ssh -J leaf1.eu.example.com:443 --cluster leaf1 node1
```

If there is no template matched, an error is returned.

```bash
$ tsh ssh -J {{proxy}} node1.none.example.com
ERROR: proxy jump contains {{proxy}} variable but did not match any of the templates in tsh config
```

> If the proxy variable `-J {{proxy}}` is not explicitly provided, then `tsh` will
> still attempt to match a template, but will not fail when there is no match.
> Additionally, it will not replace the `proxy` value when explicitly set by the
> client.

Reviewer: ```suggestion
If you don't explicitly provide the proxy variable `-J {{proxy}}`, `tsh`
still attempts to match a template, but won't fail if there isn't a match.
Additionally, `tsh` won't replace the `proxy` value if it's explicitly set by the
client.
```
